pandas/core/arrays/base.py

```diff
@@ -301,6 +319,60 @@ def unique(self):
         uniques = unique(self.astype(object))
         return self._constructor_from_sequence(uniques)
 
+    def _values_for_factorize(self):
+        """Return an array suitable for factorization.
+
+        Returns
+        -------
+        ndarray
+            An array suitable for factoraization. This should maintain order
+            and be a supported dtype.
+
+        """
+        return self.astype(object)
+
+    def factorize(self, na_sentinel=-1):
+        """Encode the extension array as an enumerated type.
+
+        Parameters
+        ----------
+        na_sentinel : int, default -1
+            Value to use in the `labels` array to indicate missing values.
+
+        Returns
+        -------
+        labels : ndarray
+            An interger NumPy array that's an indexer into the original
+            ExtensionArray.
+        uniques : ExtensionArray
+            An ExtensionArray containing the unique values of `self`.
+
+            .. note::
+
+               uniques should *not* contain a value for the NA sentinel,
+               if values in `self` are missing.
+
+        See Also
+        --------
+        pandas.factorize : top-level factorize method that dispatches here.
+
+        Notes
+        -----
+        :meth:`pandas.factorize` offers a `sort` keyword as well.
+        """
+        # Implementor notes: There are two options for overriding the
+        # behavior of `factorize`: here and `_values_for_factorize`.
+        from pandas.core.algorithms import _factorize_array
+
+        mask = self.isna()
+        arr = self._values_for_factorize()
+        arr[mask] = np.nan
```

Comment:
shouldn't this be the ``self._na_value`` ?

Reply:
We removed that from the API. Anyway, that would be the type-specific NA value, we need an NA value that pandas knows about, which is just NaN or NaT.

Speaking of which, I don't really like how this is done right now. If your values_for_factorize are integers, this will force them to be floating. Will thing more about it.